### PR TITLE
server: precisely route cancel query/session requests for tenants

### LINKED ISSUE #75252: server: the tenant cancel query/session logic unnecessarily fans out to all pods
The code in `CancelQuery` and `CancelSession` fans out the query to all pods.

This is unnecessary, because the session ID / query ID contains the target node ID (and thus SQL instance ID) so it can be routed precisely.

See the code in `(*statusServer).CancelQuery()` for an example.

Jira issue: CRDB-12531

### PR DESCRIPTION
**Pull request by matthewtodd**
Fixes #75252.

Previously, these endpoints fanned out across all SQL instances, which
was unnecessary, because both the query ID and the session ID contain an
embedded node ID (which can be treated as a SQL instance ID).

In this change, we use that embedded ID to precisely route these
cancel requests.

Release justification: Category 2: Bug fixes and low-risk updates to new
functionality.

Release note: none

### CODE REVIEW

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @matthewtodd and @rafiss)

___
*[pkg/server/tenant_status.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjlq7l3ykzvH5e0aSN:-Mxjlq7l3ykzvH5e0aSO:bqnwqwi) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L297)):*
> ```Go
> 	ctx = t.AnnotateCtx(ctx)
> 
> 	sessionID := sql.BytesToClusterWideID(req.SessionID)
> ```

Hmm, why do we need to parse the sessionID into a uint128 here? Seems like most of the API can work directly with the raw byte slice here :thinking:

**Review by matthewtodd — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @rafiss)

___
*[pkg/server/tenant_status.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjlq7l3ykzvH5e0aSN:-Mxjo-p-8pt-2hqL4vMn:bhb0c2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L297)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm, why do we need to parse the sessionID into a uint128 here? Seems like most of the API can work directly with the raw byte slice here :thinking: 
</blockquote></details>

For the call on the following line, `sessionID.GetNodeID()`.

**Review by matthewtodd — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @rafiss)

___
*[pkg/server/tenant_status.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjlq7l3ykzvH5e0aSN:-MxjoPXF9j7iRPDgCLee:b-y0xpq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L297)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

For the call on the following line, `sessionID.GetNodeID()`.
</blockquote></details>

See [`pkg/sql/cluster_wide_id.go`](https://github.com/cockroachdb/cockroach/blob/a161490fe30aa1f5f2cde30df27880e4dbdf544a/pkg/sql/cluster_wide_id.go)

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @matthewtodd and @rafiss)

___
*[pkg/server/tenant_status.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjlq7l3ykzvH5e0aSN:-MxjpKWhAQvIq2m1K3wM:bn7serg) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L297)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

See [`pkg/sql/cluster_wide_id.go`](https://github.com/cockroachdb/cockroach/blob/a161490fe30aa1f5f2cde30df27880e4dbdf544a/pkg/sql/cluster_wide_id.go)
</blockquote></details>

Hmmm interesting. I forgot if this has been discussed before, request protobuf for this endpoint already [includes](https://github.com/cockroachdb/cockroach/blob/bbdfe480ce91a7d80d03fed1fdbe4d6eb183cd6e/pkg/server/serverpb/status.proto#L994) the `node_id` field. What's our reason to parse the sessionID here instead of using that `node_id` field?

**Review by matthewtodd — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @rafiss)

___
*[pkg/server/tenant_status.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjlq7l3ykzvH5e0aSN:-Mxjq8gi0WXDZEbT0Zi1:brf70hy) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L297)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmmm interesting. I forgot if this has been discussed before, request protobuf for this endpoint already [includes](https://github.com/cockroachdb/cockroach/blob/bbdfe480ce91a7d80d03fed1fdbe4d6eb183cd6e/pkg/server/serverpb/status.proto#L994) the `node_id` field. What's our reason to parse the sessionID here instead of using that `node_id` field? 
</blockquote></details>

As I understand it, the `node_id` field here is perhaps a mistake. The session id and the query id are exactly what we need to route these requests.

**Review by Azhng — APPROVED** [review]
:lgtm: as long as the tests pass 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-:-MxjslTFEJxdWJpPT3it:bl9dooo)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @matthewtodd and @rafiss)

___
*[pkg/server/tenant_status.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjlq7l3ykzvH5e0aSN:-MxjrNrb3LS1YSszBV4r:b-poldrm) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L297)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

As I understand it, the `node_id` field here is perhaps a mistake. The session id and the query id are exactly what we need to route these requests.
</blockquote></details>

Ah I see :+1: 
___
*[pkg/server/tenant_status.go, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjs01t9LFBVeWQWxtW:-Mxjs01t9LFBVeWQWxtX:b-oka372) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L309)):*
> ```Go
> 			// NB: not using serverError() here since the priv checker
> 			// already returns a proper gRPC error status.
> 			return nil, err
> ```

Hmm not sure if this belongs in this PR, but looking at the code for `checkCancelPrivilege`, I think there are some codepath where the returning error is not wrapped in `serverError` (e.g. when it's calling `getUserAndRole`, it directly returns the error in this case, `getUserAndRole` calls `hasAdminRole()` which return some un-wrapped internal assertion errors)

**Review by rafiss — COMMENTED** [review]
there's a failure on the `tenant` logic test
```
        expected:
        could not cancel query 00000000000000000000000000000000: query ID 00000000000000000000000000000000 not found
        got:
        pq: rpc error: code = Internal desc = An internal server error has occurred. Please check your CockroachDB logs for more details.
```

i think that means the forwarding logic needs to check for invalid query IDs 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-:-Mxk4jBj5a33F_IvJnf9:b-95kkb4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng, @matthewtodd, and @rafiss)

___
*[pkg/server/tenant_status.go, line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxk4bj89KuDYsN_nElF:-Mxk4bj89KuDYsN_nElG:b-5cs5fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L220)):*
> ```Go
> 	}
> 
> 	// This request needs to be forwarded to another node.
> ```

could the privilege checking be done before forwarding to another node?

**Review by matthewtodd — COMMENTED** [review]
Thanks, @rafiss! I tried to dig through the test logs and was having a hard time identifying the failure. I've got a proposed fix here now, and `./dev testlogic --files=tenant` passes locally.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-:-MxkZi5k3S_z1kyHSGUQ:bqb36eg)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng and @rafiss)

___
*[pkg/server/tenant_status.go, line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxk4bj89KuDYsN_nElF:-MxkY8DK1HNPZZiOisFG:bvvnt4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L220)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

could the privilege checking be done before forwarding to another node?
</blockquote></details>

AFAICT, the session registry is an in-memory list of sessions on the current node only, so the check needs to happen after forwarding.
___
*[pkg/server/tenant_status.go, line 297 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjlq7l3ykzvH5e0aSN:-MxkY7NWCRF3QbDFz3qP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L297)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Ah I see :+1: 
</blockquote></details>

Done.
___
*[pkg/server/tenant_status.go, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjs01t9LFBVeWQWxtW:-MxkYjVGDhkcdYmQn9Qh:bd8fmbk) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L309)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm not sure if this belongs in this PR, but looking at the code for `checkCancelPrivilege`, I think there are some codepath where the returning error is not wrapped in `serverError` (e.g. when it's calling `getUserAndRole`, it directly returns the error in this case, `getUserAndRole` calls `hasAdminRole()` which return some un-wrapped internal assertion errors) 
</blockquote></details>

Agreed, I'm not sure that change belongs here. I'm finding that the `serverError` wrapping keeps clients from seeing specifics of the error; but we do sometimes want to share those.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @matthewtodd, and @rafiss)

___
*[pkg/server/tenant_status.go, line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxk4bj89KuDYsN_nElF:-MxkxTCf8s0gnpznfMKR:b-d9p8bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L220)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

AFAICT, the session registry is an in-memory list of sessions on the current node only, so the check needs to happen after forwarding.
</blockquote></details>

got it that makes sense...

but before this PR, the `t.checkCancelPrivilege` call was happening before any fan out was done. so how did that work before?
___
*[pkg/server/tenant_status.go, line 222 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxkx6T_9qS2I1WQTB3A:-Mxkx6TeAW-xM2od8Jky:b-1veepn) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb56c6f461069dab4e654c097726d9853a28c27d/pkg/server/tenant_status.go#L222)):*
> ```Go
> 	// This request needs to be forwarded to another instance.
> 	// We return a user-facing error, rather than a server one, before attempting to contact this known-bad instance ID.
> 	if instanceID == 0 {
> ```

i don't follow why this is the right fix. why wasn't the `sqlInstanceProvider.GetInstance(ctx, instanceID)` line returning an error if the instanceID is invalid?

also, the queryID could contain any other invalid instanceID. why is 0 special? e.g., we'd also want a nice error if someone did `CANCEL QUERY 'completely_random_query_id';`

**Review by matthewtodd — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng and @rafiss)

___
*[pkg/server/tenant_status.go, line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxk4bj89KuDYsN_nElF:-Mxp1FDu5ht3rez-Fqtb:b-c4e7qs) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L220)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

got it that makes sense...

but before this PR, the `t.checkCancelPrivilege` call was happening before any fan out was done. so how did that work before?
</blockquote></details>

Hrm, turns out I was wrong! Thanks for your careful attention here, @rafiss. I've pulled the privilege checking up to the top again.
___
*[pkg/server/tenant_status.go, line 222 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxkx6T_9qS2I1WQTB3A:-Mxp1RViAUZo9pfRt_dh:b-5kwqmh) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb56c6f461069dab4e654c097726d9853a28c27d/pkg/server/tenant_status.go#L222)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't follow why this is the right fix. why wasn't the `sqlInstanceProvider.GetInstance(ctx, instanceID)` line returning an error if the instanceID is invalid?

also, the queryID could contain any other invalid instanceID. why is 0 special? e.g., we'd also want a nice error if someone did `CANCEL QUERY 'completely_random_query_id';`
</blockquote></details>

Yeah, this was a hack to make one existing test pass. Bad idea on my part!

So I took a step back and added, in a separate commit just before this one, tests asserting on the error messages we see when canceling "bad" query and session IDs. That test passes before the main changes here and passes, unchanged, after this work, so AFAICT, we're no longer breaking observable behavior here.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @matthewtodd, and @rafiss)

___
*[pkg/server/tenant_status.go, line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxk4bj89KuDYsN_nElF:-Mxp4iWmBnNG_v7n0LY4:b-qe0oae) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L220)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Hrm, turns out I was wrong! Thanks for your careful attention here, @rafiss. I've pulled the privilege checking up to the top again.
</blockquote></details>

ah sorry, i meant to just ask in order to be educated

i am fairly sure you are right about the session registry only containing local sessions. i wonder if the issue here is with tests. if a test is either (1) using an admin user or (2) not exercising the forwarding logic then it would miss the possible bug that is caused by checking the cancel privilege before forwarding

the reason being that the findSession call is skipped for admins: https://github.com/cockroachdb/cockroach/blob/25fe0db12911a60571c3788b9403a6f42b52e003/pkg/server/status.go#L286
___
*[pkg/server/tenant_status.go, line 222 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxkx6T_9qS2I1WQTB3A:-Mxp4edaDMXPqV5bN6-3:buz6udz) ([raw file](https://github.com/cockroachdb/cockroach/blob/eb56c6f461069dab4e654c097726d9853a28c27d/pkg/server/tenant_status.go#L222)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Yeah, this was a hack to make one existing test pass. Bad idea on my part!

So I took a step back and added, in a separate commit just before this one, tests asserting on the error messages we see when canceling "bad" query and session IDs. That test passes before the main changes here and passes, unchanged, after this work, so AFAICT, we're no longer breaking observable behavior here.
</blockquote></details>

nice, thanks for the extra tests!
___
*[pkg/ccl/serverccl/statusccl/tenant_status_test.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxp6n2m4sO7XjxmfTf9:-Mxp6n2m4sO7XjxmfTfA:bfceu68) ([raw file](https://github.com/cockroachdb/cockroach/blob/88bf38813f57c804755c20659193c24477f06f44/pkg/ccl/serverccl/statusccl/tenant_status_test.go#L91)):*
> ```Go
> 
> 	t.Run("tenant_cancel_session_error_messages", func(t *testing.T) {
> 		testTenantStatusCancelSessionErrorMessages(t, testHelper)
> ```

i wonder if we need one of these tests to use a non-admin user

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @matthewtodd, and @rafiss)

___
*[pkg/server/tenant_status.go, line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxk4bj89KuDYsN_nElF:-Mxp7H0R8ekRN0xB-6T0:bev8v5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L220)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ah sorry, i meant to just ask in order to be educated

i am fairly sure you are right about the session registry only containing local sessions. i wonder if the issue here is with tests. if a test is either (1) using an admin user or (2) not exercising the forwarding logic then it would miss the possible bug that is caused by checking the cancel privilege before forwarding

the reason being that the findSession call is skipped for admins: https://github.com/cockroachdb/cockroach/blob/25fe0db12911a60571c3788b9403a6f42b52e003/pkg/server/status.go#L286
</blockquote></details>

i'm fairly certain your first instinct was right, since the non-tenant status.go checks the privilege after forwarding

https://github.com/cockroachdb/cockroach/blob/25fe0db12911a60571c3788b9403a6f42b52e003/pkg/server/status.go#L2786

**Comment by rafiss** [comment]
added @knz as an optional reviewer - he might know more about the conventions of permission checking pre- or post-forwarding.

**Review by matthewtodd — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @knz, and @matthewtodd)

___
*[pkg/server/tenant_status.go, line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxk4bj89KuDYsN_nElF:-Mxp7nF4DeLAnq2DBr71:bybij04) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L220)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'm fairly certain your first instinct was right, since the non-tenant status.go checks the privilege after forwarding

https://github.com/cockroachdb/cockroach/blob/25fe0db12911a60571c3788b9403a6f42b52e003/pkg/server/status.go#L2786

</blockquote></details>

Good call, thanks. I'll add some non-admin tests.

**Review by matthewtodd — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @knz, @matthewtodd, and @rafiss)

___
*[pkg/ccl/serverccl/statusccl/tenant_status_test.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxp6n2m4sO7XjxmfTf9:-MxpZQZqCggCoGlaSZy2:b-64xgsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/88bf38813f57c804755c20659193c24477f06f44/pkg/ccl/serverccl/statusccl/tenant_status_test.go#L91)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i wonder if we need one of these tests to use a non-admin user
</blockquote></details>

Good idea, I've added error coverage for non-admin users. It turns out we were returning opaque internal server errors for them, and that behavior's so far unchanged by this PR. I suspect we'll want to return nicer messages, though.

Before I do that, I'll add (slash make sure we have) tests that show non-admin users can indeed cancel their own queries and sessions across tenants. (Our other conversations about the in-memory session registry make me feel shaky about the test coverage here. What may have been saving us is cockroach cloud generally just spinning up one sql instance per tenant.)

**Review by knz — COMMENTED** [review]
no objection on the overall structure. Just a Q on username normalization.

> Reviewed 3 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-:-Mxsxmhk0qaBgRVlH9eK:b-42v1w7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @matthewtodd, and @rafiss)

___
*[pkg/server/tenant_status.go, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjs01t9LFBVeWQWxtW:-Mxswuq0DF9oZvE5BlJ_:bi9ltst) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L309)):*
Archer is right though. The checkCancelPrivilege function looks incomplete in that regard. Can you add a commit to fix this, or file a folllowup PR?

> I'm finding that the serverError wrapping keeps clients from seeing specifics of the error

This is a known problem and its solution deserves an independent PR.
___
*[pkg/server/tenant_status.go, line 194 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-MxswLZg-2cRJJ6rMU35:-MxswLZg-2cRJJ6rMU36:b-t9rla7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b255784dec683e912c0ec93bbef12233943a5a7c/pkg/server/tenant_status.go#L194)):*
> ```Go
> 	ctx = t.AnnotateCtx(ctx)
> 
> 	reqUsername, err := security.MakeSQLUsernameFromPreNormalizedStringChecked(req.Username)
> ```

Wait why did this need to change? I don't see how the username can be invalid at this point.
___
*[pkg/server/tenant_status.go, line 305 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxswpnm9Z_JfUUpgTU5:-Mxswpnm9Z_JfUUpgTU6:b-3nokg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b255784dec683e912c0ec93bbef12233943a5a7c/pkg/server/tenant_status.go#L305)):*
> ```Go
> 	ctx = t.AnnotateCtx(ctx)
> 
> 	reqUsername, err := security.MakeSQLUsernameFromPreNormalizedStringChecked(req.Username)
> ```

same question as above.

**Review by matthewtodd — COMMENTED** [review]
Thanks, everyone. I think we're at a closure point here now: I've fixed the username normalization, filed #77676 for the checkCancelPrivilege work, and will be sending a separate PR with tests around non-admin cancellation behavior.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-:-MxtdQXC680a_qWqQHua:b-5qr1um)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @knz, @matthewtodd, and @rafiss)

___
*[pkg/server/tenant_status.go, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxjs01t9LFBVeWQWxtW:-Mxta6TuDG57fKgcY6FW:b30asfm) ([raw file](https://github.com/cockroachdb/cockroach/blob/748b48d270877682137a176d1c612e4f3217778e/pkg/server/tenant_status.go#L309)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Archer is right though. The checkCancelPrivilege function looks incomplete in that regard. Can you add a commit to fix this, or file a folllowup PR?

> I'm finding that the serverError wrapping keeps clients from seeing specifics of the error

This is a known problem and its solution deserves an independent PR.
</blockquote></details>

Thanks, filed the checkCancelPrivilege follow up work in #77676.
___
*[pkg/server/tenant_status.go, line 194 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-MxswLZg-2cRJJ6rMU35:-Mxt_bem2cTjJ3nDKfFd:bc5qiks) ([raw file](https://github.com/cockroachdb/cockroach/blob/b255784dec683e912c0ec93bbef12233943a5a7c/pkg/server/tenant_status.go#L194)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Wait why did this need to change? I don't see how the username can be invalid at this point.
</blockquote></details>

Ah, good to know. This was just me thinking "error checking is good" without understanding the surrounding context. Reverted.
___
*[pkg/server/tenant_status.go, line 305 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxswpnm9Z_JfUUpgTU5:-Mxt_u_EEiKpNHccfja9:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b255784dec683e912c0ec93bbef12233943a5a7c/pkg/server/tenant_status.go#L305)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

same question as above.
</blockquote></details>

Fixed.
___
*[pkg/ccl/serverccl/statusccl/tenant_status_test.go, line 91 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-Mxp6n2m4sO7XjxmfTf9:-Mxtcm9D-gEVqXE1b0P-:b-mgdrjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/88bf38813f57c804755c20659193c24477f06f44/pkg/ccl/serverccl/statusccl/tenant_status_test.go#L91)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Good idea, I've added error coverage for non-admin users. It turns out we were returning opaque internal server errors for them, and that behavior's so far unchanged by this PR. I suspect we'll want to return nicer messages, though.

Before I do that, I'll add (slash make sure we have) tests that show non-admin users can indeed cancel their own queries and sessions across tenants. (Our other conversations about the in-memory session registry make me feel shaky about the test coverage here. What may have been saving us is cockroach cloud generally just spinning up one sql instance per tenant.)
</blockquote></details>

Checking back in, I've worked up some tests for non-admin users canceling their own queries and sessions across sql instances, but they were a large enough change with their own open questions that I'd like to file those for a separate review. So I think this PR's good to go?

**Review by maryliag — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 3 files at r5, 1 of 1 files at r6, 1 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-:-MxtezwT4jGmOhlVPotc:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @knz, @maryliag, @matthewtodd, and @rafiss)

**Review by rafiss — APPROVED** [review]
looks like the permission checking is a pre-existing problem, so i agree with not making it block this PR. i left a nit for comments, but feel free to merge this whenever you want

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-:-MxtwTGk9mMCJiSqtX3J:biro772)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @knz, @maryliag, @matthewtodd, and @rafiss)

___
*[pkg/server/tenant_status.go, line 195 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-MxtvwJ3B14mTnwTa0vM:-MxtvwJ3B14mTnwTa0vN:btjoadn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea93f91b0a50b57bf6efe2a9e7b5da041a31abc8/pkg/server/tenant_status.go#L195)):*
> ```Go
> 
> 	reqUsername := security.MakeSQLUsernameFromPreNormalizedString(req.Username)
> 	if err := t.checkCancelPrivilege(ctx, reqUsername, findSessionByQueryID(req.QueryID)); err != nil {
> ```

nit: if you can't address the permission checking right now, i think a TODO comment that explains the problem and that links to an open issue would be very useful. (actually, even without the new tests, couldn't this pretty easily be moved to check permissions after forwarding?)
___
*[pkg/ccl/serverccl/statusccl/tenant_status_test.go, line 1015 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-MxtvZtLD1v90b4RrhIK:-MxtvZtLD1v90b4RrhIL:b-7m14e4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2c1633eefecaaeecd655fec7ab539216fb193ba/pkg/ccl/serverccl/statusccl/tenant_status_test.go#L1015)):*
> ```Go
> 	}
> 
> 	for _, isAdmin := range []bool{true, false} {
> ```

an optional nit: you could use `testutils.RunTrueAndFalse`

**Review by matthewtodd — COMMENTED** [review]
Thanks! Tidied up the nits and will merge on green.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-:-My7vL-wEqiKkW7oFbft:bhc6kbv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @Azhng, @knz, @maryliag, @matthewtodd, and @rafiss)

___
*[pkg/server/tenant_status.go, line 195 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-MxtvwJ3B14mTnwTa0vM:-My7oxUO5urJcpXJraUr:bsf1k4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea93f91b0a50b57bf6efe2a9e7b5da041a31abc8/pkg/server/tenant_status.go#L195)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: if you can't address the permission checking right now, i think a TODO comment that explains the problem and that links to an open issue would be very useful. (actually, even without the new tests, couldn't this pretty easily be moved to check permissions after forwarding?)
</blockquote></details>

Agreed, moved the permissions check down, linking to #77676 in the tests.
___
*[pkg/ccl/serverccl/statusccl/tenant_status_test.go, line 1015 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/77550#-MxtvZtLD1v90b4RrhIK:-My7v2S41Kf9yU-2DJ81:bpe33mm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2c1633eefecaaeecd655fec7ab539216fb193ba/pkg/ccl/serverccl/statusccl/tenant_status_test.go#L1015)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

an optional nit: you could use `testutils.RunTrueAndFalse`
</blockquote></details>

Oh, nice! Done.

**Comment by matthewtodd** [comment]
There will be a small merge conflict here when #77732 lands (it's already in the bors queue). I've got a patch ready to go.

**Pull request merged**
